### PR TITLE
Sort out argv emulation for macOS app bundles

### LINKED ISSUE #5436: [proposed area:bootloader platform:OS X] Mac OS X Application Does Not Receive Drag 'n Drop File Paths
A Mac OS X application built with PyInstaller 4.1 `--onedir` and `--windowed` is run as single process (cf., a forked child process). Consequently, when it is started by a drag and drop operation, it does not process `OpenDoc` Apple Events and `sys.argv` does not include the dropped path(s).

From `pyi_main.c:pyi_main ()`:
```
#if defined(_WIN32) || defined(__APPLE__)

    /* On Windows and Mac use single-process for --onedir mode. */
    if (!extractionpath && !pyi_launch_need_to_extract_binaries(archive_status)) {
        VS("LOADER: No need to extract files to run; setting extractionpath to homepath\n");
        extractionpath = homepath;
    }

#endif
    :
    :
    if (extractionpath) {
        VS("LOADER: Already in the child - running user's code.\n");
          :
          :
    }
    else {
          :
          :
        /* Run the 'child' process, then clean up. */

        VS("LOADER: Executing self as child\n");
          :
          :
    }
```

### LINKED ISSUE #5908: File path aruguments not received on macOS
## Description of the issue

The app built successfully, however, when I right-click on a file and choose this app to open it, no arguments are received in python script.

However, it works when I manually pass the arguments in Terminal by `xxx.app/Contents/MacOS/executable Arg1 Arg2`, maybe the `kAEOpenDocuments` event is not handled correctly?

### Context information (for bug reports)

* Output of `pyinstaller --version`: ```4.3```
* Version of Python: 3.9.5
* Platform: macOS 11.4
* How you installed Python: pyenv
* Did you also try this on another platform? Does it work there?
This is a macOS-specific issue.

### A minimal example program which shows the error
Built a test app using the script & spec file below, right-click on any file and choose `test.app` to open.

document_event_test.py
```
#!/usr/bin/env python3
#encoding:utf-8

import sys, os

base = os.path.dirname(sys.argv[0])

with open(os.path.join(base, "logs.txt"), "w+") as f:
    if len(sys.argv) > 1:
        f.write("arguments received:\n")
        for arg in sys.argv[1:]:
            f.write(f"{arg}, is file: {os.path.isfile(arg)}\n")
    else:
        f.write("no arguments received!\n")

```

spec file:
```
# -*- mode: python ; coding: utf-8 -*-

block_cipher = None

a = Analysis(['document_event_test.py'],
             pathex=[],
             binaries=[],
             datas=[],
             hiddenimports=[],
             hookspath=[],
             runtime_hooks=[],
             excludes=[],
             win_no_prefer_redirects=False,
             win_private_assemblies=False,
             cipher=block_cipher,
             noarchive=False)

pyz = PYZ(a.pure, a.zipped_data,
             cipher=block_cipher)

exe = EXE(pyz,
          a.scripts,
          [],
          exclude_binaries=True,
          name='test',
          debug=False,
          bootloader_ignore_signals=False,
          strip=False,
          upx=False,
          console=False, 
	 icon='')

coll = COLLECT(exe,
               a.binaries,
               a.zipfiles,
               a.datas,
               strip=False,
               upx=False,
               upx_exclude=[],
               name='test')

app = BUNDLE(coll,
            name='test.app',
            bundle_identifier="test",
            info_plist={
                'NSPrincipalClass': 'NSApplication',
                'NSAppleScriptEnabled': False,
                'CFBundleDocumentTypes': [{
                        'CFBundleTypeName': 'test_file_types',
                        'CFBundleTypeExtensions': ['txt', 'std', 'stdf', 'std*'],
                        'CFBundleTypeRole': 'Viewer'
                        }],
                },
            )
```

### Stacktrace / full error message
A log file is generated in the example above, but the content of the log is always `no arguments received!`.

### PR DESCRIPTION
**Pull request by rokm**
This is a revised attempt at implementing *argv emulation* for `onedir` app bundles. This time, *argv emulation* is implemented as an opt-in feature instead of forced built-in behavior, which allows user to avoid potential conflict with UI toolkits in `onedir` mode.

The feature is now also documented, along with a warning about using it in combination with UI toolkits in `onedir` mode.

Closes #5436, closes #5908; hopefully for good this time.

To keep behavior consistent between `onefile` and `onedir` bundles, the *argv emulation* is made opt-in for both modes, whereas previously, it was forced built-in behavior for `onefile` bundles. Therefore, the change is marked as a breaking one, although I would argue that this should have been an option rather than built-in behavior from the very beginning. For example, GUI applications need to implement proper event handlers to handle (potential) subsequent events anyway, and forced *argv emulation* offers them no benefit (if anything, it forces them to consider sys.argv in addition to implementing event handlers...).

For `onedir` bundles with *argv emulation* enabled, I've added a work-around that attempts to mitigate the Tk issues from #6043. But officially, we should still discourage the use of *argv emulation* in combination with UI frameworks, same as py2app does.

The majority of PR consists of step-by-step refactoring of the Apple Event handling code and revising its API into more explicit functions, such as separate handler installation step, a function for handler uninstallation/cleanup, etc. I've also moved the Apple Event handling code into separate source file to make it easier to navigate.

### CODE REVIEW

**Comment by rokm** [comment]
Alright, marking this ready for review.

**Comment by bwoodsend** [comment]
Ughh, this is way over my head. All I can suggest is that it would be nice to see some hello world code examples of receiving an event for some of the permutations listed in b1046922be1eb3b7e4cc1798390fd83fb74fa0bd. And there's a few ``` ``--argv-emulation`` ```s which could be cross references. Other than that, I'll have to take CI's word for it that this works...

**Comment by rokm** [comment]
I've added cross-references for those `--argv-emulation` occurrences.

I need to think about the event handling examples, because this is not quite hello-world-level functionality (it depends on the UI toolkit, plus you need to register file extensions/URL schemas in the .spec file). I suppose I could repurpose my tkinter- and PySide2-based event loggers that didn't make it into the automated test into exemplar GUI loggers for open doc/url events instead...

**Comment by bwoodsend** [comment]
If I understand [this](https://github.com/pyinstaller/pyinstaller/pull/6089/commits/71dbaead84ddb393905ce12b933cf2a3e8c63155#diff-ed2ef29371b4d95c70034a5aa21c4cc75e68b3ce5740ef6ee973f171e916d583R316-R326) correctly, with `argv_emulation` enabled, an application only has to read `sys.argv` and it will be launched to open files of an associated type? That option doesn't sound like more than a few lines to demonstrate? As for the other stuff, maybe leave it then. Is this event handling possible in regular unfrozen Python? i.e. Will there be other sources of examples in the various GUI toolkits' docs?

**Comment by rokm** [comment]
> That option doesn't sound like more than a few lines to demonstrate?

Sure, but the "code" part itself is trivial (i.e., read ``sys.argv``). It's everything else (like setting up associations in .spec file) that is a hassle here... But I think I came up with decent example for an application using argv emu.

> Is this event handling possible in regular unfrozen Python?

I'm not sure if it is. I mean, it is probably possible to set up a bundle with bash launcher script that runs a python script in system-wide python interpreter, but I have no idea how practical and wide-spread it is.

>  i.e. Will there be other sources of examples in the various GUI toolkits' docs?

There is some documentation available for [Tcl/Tk](https://www.tcl.tk/man/tcl8.5/TkCmd/tk_mac.html) and [Qt](https://doc.qt.io/qt-5/qfileopenevent.html), but I think it again tells only a part of the story. 

I suppose for people not very familiar with the topic, it's actually indeed best to have a complete example.

**Review by bwoodsend — CHANGES_REQUESTED** [review]
Whoah that's more than I expected. I don't think PyInstaller has ever seen that much documentation! There's a couple of missing `the`s but other than that it looks great.

File: doc/feature-notes.rst [hunk 1]
```diff
@@ -346,6 +346,484 @@ mode.) and as such need to be handled via event handlers.
...
+           }],
+       }
+   )
+
+
+Open event handling with argv emulation
+---------------------------------------
+
+Consider the following python script that began its life as a command-line
+utility, to be evoked from terminal as follows::
```
[the remark this answers is not in the document]
Comment (rokm) [reply doc/feature-notes.rst]: Indeed.

**Review by bwoodsend — APPROVED** [review]

**Comment by jamesbcd** [comment]
This is a really well-thought-out design, and the documentation is extraordinarily good. Thanks so much for developing it.

**Pull request merged**